README.md
<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/headlines/ascii-to-svg.svg" height="60px" banner="Ascii Art To Svg">

**Ascii to Svg Generator for Ruby**

This module generates beautiful svg images based on a input string.

## Examples

<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/banners/0.svg" headline="Example">

```ruby
    character_set = [ '-', '/', '|', "\", '#' ]
```

<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/banners/3.svg">

```ruby
    character_set = [ '-', '|', '#' ]
```

<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/banners/4.svg">

```ruby
    character_set = [ '-', '|', '#', 'o', '\' ]
```

<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/banners/6.svg">

```ruby
    character_set = [ '|', 'o', '.' ]
    options = {
        style__line__stroke__color: 'brown',
        style__ellipse__stroke__color: 'orange'
    }
```

<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/banners/7.svg">

```ruby
    character_set = [ '/', '\' ]
    options = {
        style__line__stroke__color: 'brown',
        style__ellipse__stroke__color: 'orange'
    }
```

<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/banners/12.svg">

```ruby
    character_set = [ 'x', '.' ]
```

<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/banners/14.svg">

```ruby
    character_set = [ '|', '.', '-' ]
    options = {
        style__canvas__fill__color: '#0A0C10',
        style__line__stroke__color: '#8D949D'
    }
```

<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/headlines/quickstart.svg" height="60px" banner="Quickstart">

1. Install `gem`
```ruby
    gem install 'ascii_to_svg'
```

2. Run Code
```ruby
    require 'ascii_to_svg'

    # Generate Example String
    ascii = AsciiToSvg.example_string( [ 'x', 'o' ], 256 )

    # Generate SVG
    svg = AsciiToSvg.from_string( ascii, 16, {} )

    # => "<svg xmlns=\"http://www.w3.org/2000/svg ...
```

<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/headlines/functions.svg" height="60px" banner="Functions">

### AsciiToSvg.options<br>
  Output all Parameter
```ruby
    AsciiToSvg.options
    # => {:canvas=>{:size=>{:x=>500 ...
```
### AsciiToSvg.example_string()

| **Type** | **Required** | **Description** | **Example** | **Description** |
|------:|:------|:------|:------|:------| 
| **characters** | ```Array (of single alphabetic character)``` | No | ```[ 'x', 'o' ]``` | Set alphabetic characters for string |
| **length** | ```Integer``` | No | ```512``` | Lengt of generated String | |
<br>

```ruby
    one = AsciiToSvg.example_string()
    two = AsciiToSvg.example_string( [ 'x', 'o' ], 512 )
```

### AsciiToSvg.from_string()

| | **Type** | **Required** | **Description** |
|------:|:------|:------|:------|
| **ascii** | ```String``` | Yes | A String of one or more Characters |
| **lenght** | ```Integer``` | Yes | Number of Characters in one Line (Row) |
| **options** | ```Hash``` |  No | Modify output. For detailed Information | 
<br>

```ruby
    AsciiToSvg.from_string(
        ascii, 
        length, 
        options
    )
```


### AsciiToSvg.similar_svg()

```ruby
    original_str = AsciiToSvg.example_string( [ 'x', 'o' ], 256 )
    original_svg = AsciiToSvg.from_string( original_str, 16, {} )

    copy_str = original_str[ 0, original_str.length - 1 ]
    copy_svg = AsciiToSvg.from_string( copy_str, 16, {} )

    AsciiToSvg.similar_svg( original_svg, copy_svg )
    # => => {:hexdigest1=>"79d5e81d230214749672a1c10e103c46", :hexdigest2=>"f48cacecf2720633dd081c3421d84981", :unique=>false, :score=>0.002799275481640046} 
```  

<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/headlines/naming.svg" height="60px" banner="Naming">

```
CANVAS
-------------------------------------
|                   ^
|                   | Margin Top
|        GRID       v
|        ----------------------------
|        |   CELL
|        |   ------------------------
|        |   |           | Offset x |
| Margin |   |   Symbol  |<-------->|
| Left   |   |           |          |
| <----> |   |------------          |
|        |   |        ^             |
|        |   |        | Offset Y    |
|        |   |        V             |
|        |   ------------------------
```

<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/headlines/symbols.svg" height="60px" banner="Symbols">

| Nr | Sign | Image | SVG Element |
| :-- | :-- | :-- | :-- |
| 1 | "\\" | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/3-tl-br.svg"> | Line |
| 2 | "\|" | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/2-vertical.svg"> | Line |
| 3 | "/" | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/1-tr-bl.svg"> | Line |
| 4 | "-" | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/0-minus.svg"> | Line |
| 5 | "+" | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/4-plus.svg"> | Line |
| 6 | "x" | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/7-x.svg"> | Line |
| 7 | "o" | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/8-ellipse.svg"> | Ellipse |
| 8 | "#" | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/5-rectangle.svg"> | Rectangle |
| 9 | "." | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/6-empty.svg"> | |



<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/headlines/options.svg" height="60px" banner="Options">

### Canvas
Defines the full width and height.
| Nr | Name | Key | Default | Type | Description |
| :-- | :-- | :-- | :-- | :-- | :-- |
| A.1. | Size X |:canvas__size__x | `500` | Integer | Set width of canvas |
| A.2. | Margin Left |:canvas__margin__left | `0` | Integer | Set margin on the left side |
| A.3. | Margin Top |:canvas__margin__top | `0` | Integer | Set margin on the top |
| A.4. | Margin Right |:canvas__margin__right | `0` | Integer | Set margin on the right |
| A.5. | Margin Bottom |:canvas__margin__bottom | `0` | Integer | Set margin on the bottom |


### Cell
Defines the Area of one symbol including offset.
| Nr | Name | Key | Default | Type | Description |
| :-- | :-- | :-- | :-- | :-- | :-- |
| B.1. | X Offset |:cell__x__offset | `0` | Integer | Define the X (width) offset to the next cell |
| B.2. | Y Offset |:cell__y__offset | `0` | Integer | Define the Y (height) offset to the next cell |


### Symbols
Defines which `char` will be interpreted as "`svg element`"
| Nr | Name | Key | Default | Type | Image |
| :-- | :-- | :-- | :-- | :-- | :-- |
| C.1. | \ |:"symbols__\\\\" | `[ "\\" ]` | Array | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/3-tl-br.svg"> |
| C.2. | / | :"symbols__/" | `[ "/" ]` | Array | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/1-tr-bl.svg"> |
| C.3. | X | :"symbols__X" | `[ "X", "x" ]` | Array | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/7-x.svg"> |
| C.4. | - | :"symbols__-" | `[ "-" ]` | Array | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/0-minus.svg"> |
| C.5. | \| | :"symbols__\" | | `[ "\|", "1" ]` | Array | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/2-vertical.svg"> |
| C.6. | O | :"symbols__O" | `[ "O", "o", "0" ]` | Array | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/8-ellipse.svg"> |
| C.7. | + | :"symbols__+" | `[ "+" ]` | Array | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/4-plus.svg"> |
| C.8. | # | :"symbols__#" | `[ "#" ]` | Array | <img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/symbols/5-rectangle.svg"> |


### Style
Defines all Style Attributes. Styles can only changed by type of the svg element except "Canvas". Under "Symbol" you can find out which Symbol uses which svg Element.

[**Line**](#line)
| Nr | Name | Key | Default | Type | Description |
| :-- | :-- | :-- | :-- | :-- | :-- |
| D.1. | Stroke Width |:style__line__stroke__width | `2.0` | Float | Define width of stroke, please notice linecap type for desired behavior |
| D.2. | Stroke Color |:style__line__stroke__color | `"rgb(0,0,0)"` | String | Define color of stroke in RGB, you can also use HTML Color names or "none" |
| D.3. | Stroke Opacity |:style__line__stroke__opacity | `1.0` | Float | Define opacity of the stroke, use floating numbers. |
| D.4. | Stroke Linecap |:style__line__stroke__linecap | `"square"` | String | Defines behavior of line ("butt" / "round" / "sqaure"). Detailed explaination: [developer.mozilla.org](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/stroke-linecap)|

[**Ellipse**](#ellipse)
| Nr | Name | Key | Default | Type | Description |
| :-- | :-- | :-- | :-- | :-- | :-- |
| D.5. | Stroke Width |:style__ellipse__stroke__width | `2.0` | Float | Define stroke width, use floating numbers |
| D.6. | Stroke Color |:style__ellipse__stroke__color | `"rgb(0,0,0)"` | String | Define stroke color in RGB, you can also use HTML Color names or "none". |
| D.7. | Stroke Opacity |:style__ellipse__stroke__opacity | `1.0` | Float | Set stroke opacity, use floating numbers. |
| D.8. | Stroke Linecap |:style__ellipse__stroke__linecap | `"square"` | String | Defines behavior of ellipse ("butt" / "round" / "sqaure"). Detailed explaination: [developer.mozilla.org](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/stroke-linecap)| |
| D.9. | Fill |:style__ellipse__fill | `"none"` | String | Define color fill of |

[**Rectangle**](#rectangle)
| Nr | Name | Key | Default | Type | Description |
| :-- | :-- | :-- | :-- | :-- | :-- |
| D.10. | Fill Color |:style__rectangle__fill__color | `"rgb(0,0,0)"` | String | Define infill color in RGB, you can also use HTML Color names or "none". |
| D.11. | Fill Opacity |:style__rectangle__fill__opacity | `1.0` | Float | Set infill opacity, use floating numbers. |

[**Canvas**](#canvas)
| Nr | Name | Key | Default | Type | Description |
| :-- | :-- | :-- | :-- | :-- | :-- |
| D.12. | Fill Color |:style__canvas__fill__color | `"rgb(255,255,255)"` | String | Define canvas infill color in RGB, you can also use HTML Color names or "none". |
| D.13. | Fill Opacity |:style__canvas__fill__opacity | `1.0` | Float | Set canvas infill opacity, use floating numbers.|


<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/headlines/contributing.svg" height="60px" banner="Contributing">

Bug reports and pull requests are welcome on GitHub at https://github.com/a6b8/ascii_to_svg. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/a6b8/ascii_to_svg/blob/master/CODE_OF_CONDUCT.md).


<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/headlines/license.svg" height="60px" banner="License">

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


<img src="https://raw.githubusercontent.com/a6b8/a6b8/main/docs/ascii-to-svg-for-ruby/readme/headlines/code-of-conduct.svg" height="60px" banner="Code of Conduct">

Everyone interacting in the AsciiToSvg project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/a6b8/ascii_to_svg/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [Unreleased]

## [0.1.0] - 2021-06-14

- Initial release


test/t.rb
require '../lib/ascii_to_svg'

AsciiToSvg.from_string( "#.#.#", 5, {})

test/test.rb
require '../lib/ascii_to_svg'

sets = [
    {
        test: 'Draw symbols',
        characters: [ '-', '/', '|', "\\", '+' ],
        times: 512,
        chars: 32,
        path: '0-many-characters.svg',
        params: {}
    },
    {
        test: 'Change default canvas size',
        characters: [ '-', '/', '|', "\\", '+', '#', '.' ],
        times: 512,
        chars: 16,
        path: '1-many-characters.svg',
        params: { canvas__size__x: 200 }
    },
    {
        test: 'Draw squares',
        characters: [ '#' ],
        times: 256,
        chars: 16,
        path: '2-squuares.svg',
        params: {}
    },
    {
        test: 'Draw squares with offset',
        characters: [ 'o' ],
        times: 256,
        chars: 16,
        path: '3-offset.svg',
        params: { cell__x__offset: 6, cell__y__offset: 6, }
    },
    {
        test: 'Draw squares with offset',
        characters: [ 'x' ],
        times: 256,
        chars: 16,
        path: '3-offset-2.svg',
        params: { canvas__size__x: 300, cell__x__offset: 8, cell__y__offset: 8, }
    },
    {   
        test: 'Test minus cell offset values',
        characters: [ 'X', '#' ],
        times: 256,
        chars: 16,
        path: '4-offset-minus.svg',
        params: { cell__x__offset: -2, cell__y__offset: -2 }
    },
    {   
        test: 'Color and Opacity Test',
        characters: [ '/', "#", "o" ],
        times: 256,
        chars: 16,
        path: '5-color.svg',
        params: { 
            style__line__stroke__color: 'rgb(50,100,150)', 
            style__ellipse__stroke__color: 'rgb(150,200,250)',
            style__rectangle__fill__color: 'rgb(250,300,50)',
            style__rectangle__fill__opacity: 0.2
        }
    },
    {   
        test: 'Banner',
        characters: [ '-', '/', '|', "\\", '+' ],
        times: 128,
        chars: 16,
        path: '7-banner.svg',
        params: { cell__x__offset: -2, cell__y__offset: -2 }
    }
] 

puts 'TEST SVG'
for index in 0..sets.length-1
    set = sets[ index ]
    puts " [#{index}]\t" + set[:test]
    ascii = AsciiToSvg.example_string( set[:characters], set[:times] )
    svg = AsciiToSvg.from_string( ascii, set[:chars], set[:params] )
    File.open( './examples/' + set[:path], "w" ) { | f | f.write( svg ) }
end




test/example.rb
require 'ascii_to_svg'

# Generate Example String
ascii = AsciiToSvg.example_string( ['x', 'o'], 256 )

# Generate SVG
svg = AsciiToSvg.from_string( ascii, 16, {} )

# => 

lib/ascii_to_svg.rb
# frozen_string_literal: true

require_relative "ascii_to_svg/version"
require 'digest'


module AsciiToSvg
  class Error < StandardError; end

  @TEMPLATE = {
      canvas: {
          size:{
            x: 500,
            y: nil,
          },
          margin: {
              left: 0,
              top: 0,
              right: 0,
              bottom: 0
          }
      },
      grid: {
          x: {
              offset: nil,
              length: nil
          },
          y: {
              offset: nil,
              length: nil
          },
          size: {
              x: nil,
              y: nil
          }
      },
      cell: {
          x: {
            offset: 0
          },
          y: {
            offset: 0
          },
          size: {
            x: nil,
            y: nil
          }
      },
      symbols: {
        "\\": [ "\\" ],
        "/": [ "/" ],
        "X": [ "X", "x" ],
        "-": [ "-" ],
        "|": [ "|", "1" ],
        "O": [ "O", "o", "0" ],
        "+": [ "+" ],
        "#": [ "#" ]
      },
      style: {
        line: {
          stroke: {
            width: 2.0,
            color: 'rgb(0,0,0)',
            opacity: 1.0,
            linecap: 'square' # butt / round / square
          }
        },
        ellipse: {
            stroke: {
              width: 2.0,
              color: 'rgb(0,0,0)',
              opacity: 1.0,
              linecap: 'square'
            },
            fill: 'none'
        },
        rectangle: {
          fill: {
            color: 'rgb(0,0,0)',
            opacity: 1.0
          }
        },
        canvas: {
          fill: {
            color: 'rgb(255,255,255)',
            opacity: 1.0
          }          
        }
      }
  }

  def self.get_options()
    return @TEMPLATE
  end


  def self.example_string( characters=[ '-', '/', '|', "\\", '+' ], t=512 )
    return t.times
      .map { | s | characters[ rand( characters.length ) ] }
      .join()
  end


  def self.from_string( ascii, _length, vars = {} )
    if self.options_update( vars, true )
      options, lines = self.options_prepare( ascii, _length, vars )
      elements = ''
      for y in 0..options[:grid][:y][:length] - 1
        line = lines[ y ]
        chars = line.split( '' )
        for x in 0..options[:grid][:x][:length] - 1
          char = line[ x ]
          symbols = self.cell_symbols( char, options )
          position = self.cell_position( x, y, options )
          elements += self.cell_svg( symbols, position, options )      
        end
      end
      result = self.generate( elements, options )
    else
    end
  end


  def self.similar_svg( a, b )
    compare = {}

    [ [ :a, a ], [ :b, b ] ].each do | str |
      str[ 1 ].gsub!( "\n", '' )
      str[ 1 ].gsub!( ' ', '' )
      compare[ str[ 0 ] ] = str[ 1 ]
    end

    result = {
      hexdigest1: nil,
      hexdigest2: nil,
      unique: nil,
      score: nil
    }

    result[:hexdigest1] = self.get_hexdigest( compare[:a] )
    result[:hexdigest2] = self.get_hexdigest( compare[:b] )
    result[:unique] = result[:hexdigest1] == result[:hexdigest2]
    result[:score] = self.str_difference( compare[:a], compare[:b] )
    return result
  end


  def self.get_hexdigest( str )
    Digest::MD5.hexdigest str
  end


  private

  def self.str_difference( a, b )
    a = a.to_s.downcase.split( '_' ).join( '' )
    b = b.to_s.downcase.split( '_' ).join( '' )
    longer = [ a.size, b.size ].max
    same = a
      .each_char
      .zip( b.each_char )
      .select { | a, b | a == b }
      .size
    ( longer - same ) / a.size.to_f
  end


  def self.options_update( vars, validation )
    allow_list = [
      :canvas__size__x,
      :canvas__margin__left,
      :canvas__margin__top,
      :canvas__margin__right,
      :canvas__margin__bottom,
      :cell__x__offset,
      :cell__y__offset,
      :"symbols__\\",
      :"symbols__/",
      :"symbols__X",
      :"symbols__-",
      :"symbols__|",
      :"symbols__O",
      :"symbols__+",
      :"symbols__#",
      :style__line__stroke__width,
      :style__line__stroke__color,
      :style__line__stroke__opacity,
      :style__line__stroke__linecap,
      :style__ellipse__stroke__width,
      :style__ellipse__stroke__color,
      :style__ellipse__stroke__opacity,
      :style__ellipse__stroke__linecap,
      :style__ellipse__fill,
      :style__rectangle__fill__color,
      :style__rectangle__fill__opacity,
      :style__canvas__fill__color,
      :style__canvas__fill__opacity
    ]
  
    messages = []
    _options = Marshal.load( Marshal.dump( @TEMPLATE ) )
    
    vars.keys.each do | key |
      if allow_list.include?( key ) 
  
        keys = key.to_s.split( '__' ).map { | a | a.to_sym }
        case( keys.length )
          when 1
            _options[ keys[ 0 ] ] = vars[ key ]
          when 2
            _options[ keys[ 0 ] ][ keys[ 1 ] ] = vars[ key ]
          when 3
            _options[ keys[ 0 ] ][ keys[ 1 ] ][ keys[ 2 ] ] = vars[ key ]
          when 4
            _options[ keys[ 0 ] ][ keys[ 1 ] ][ keys[ 2 ] ][ keys[ 3 ] ] = vars[ key ]
        end
      else
        nearest = allow_list
          .map { | word | { score: self.str_difference( key, word ), word: word } }
          .min_by { | item | item[:score] }

        message = "\"#{key}\" is not a valid key, did you mean \"<--similar-->\"?"
        message = message.gsub( '<--similar-->', nearest[:word].to_s )
        messages.push( message )
      end
    end
    
    if messages.length != 0
      messages.length == 1 ? puts( 'Error found:' ) : puts( 'Errors found:' ) 
      messages.each { | m | puts( '- ' + m ) }
    end
    return validation ? messages.length == 0 : _options
  end


  def self.options_prepare( ascii, _length, vars )   
    options = self.options_update( vars, false )

    options[:grid][:x][:length] = _length
    
    lines = ascii.chars.each_slice( options[:grid][:x][:length] ).map( &:join )
    options[:grid][:y][:length] = lines.length

    options[:grid][:x][:offset] = options[:canvas][:margin][:left] + options[:cell][:x][:offset] / 2
    options[:grid][:y][:offset] = options[:canvas][:margin][:top] + options[:cell][:x][:offset] / 2

    tmp = [ :left, :right ].map{ | k | options[:canvas][:margin][ k ] }.sum
    options[:grid][:size][:x] = options[:canvas][:size][:x] - tmp

    tmp = ( options[:cell][:x][:offset] * options[:grid][:x][:length] )
    options[:cell][:size][:x] = ( options[:grid][:size][:x] - tmp ) / options[:grid][:x][:length]
    
    tmp = [ :top, :bottom ].map{ | k | options[:canvas][:margin][ k ] }.sum
    options[:canvas][:size][:y] = ( ( options[:cell][:size][:x] + options[:cell][:x][:offset])* options[:grid][:y][:length] ) + tmp
    options[:grid][:size][:y] = options[:canvas][:size][:y] - tmp
    
    tmp = ( options[:cell][:y][:offset] * (options[:grid][:y][:length] ) )
    options[:cell][:size][:y] = ( options[:grid][:size][:y] - tmp ) / options[:grid][:y][:length]

    return [ options, lines ]
  end


  def self.cell_symbols( char, options )
    selector = nil
    options[:symbols].keys.each do | key |
       options[:symbols][ key ].include?( char ) ? selector = key : ''
    end

    results = []
    case( selector )
      when :"\\"
        results.push( [ 'line', :top__left, :bottom__right ] )
      when :"/"
        results.push( [ 'line', :top__right, :bottom__left ] )
      when :"X"
        results.push( [ 'line', :top__right, :bottom__left ] )
        results.push( [ 'line', :top__left, :bottom__right ] )
      when :"-"
        results.push( [ 'line', :center__left, :center__right ] )
      when :"|"
        results.push( [ 'line', :top__center, :bottom__center ] )
      when :"+"
        results.push( [ 'line', :center__left, :center__right ] )
        results.push( [ 'line', :top__center, :bottom__center ] )
      when :"O"
        results.push( [ 'ellipse', :center__center, nil ] )
      when :"#"
        results.push( [ 'rectangle', :top__left, nil ] )
    end
    return results
  end


  def self.cell_position( x, y, options )
    pos = {
      top:{
        left: nil,
        center: nil,
        right: nil
      },
      center: {
        left: nil,
        center: nil,
        right: nil,
      },
      bottom: {
        left: nil,
        center: nil,
        right: nil
      }
    }

    x = ( x * ( options[:cell][:size][:x] + options[:cell][:x][:offset] ) ) + options[:grid][:x][:offset]
    y = ( y * ( options[:cell][:size][:y] + options[:cell][:y][:offset] ) ) + options[:grid][:y][:offset]

    half_x = options[:cell][:size][:x] / 2
    half_y = options[:cell][:size][:y] / 2

    pos[:top][:left] = [ x, y ]
    pos[:top][:center] = [ x + half_x, y ]
    pos[:top][:right] = [ x + options[:cell][:size][:x], y ]

    pos[:center][:left] = [ x, y + half_y ]
    pos[:center][:center] = [ x + half_x, y + half_y ]
    pos[:center][:right] = [ x + options[:cell][:size][:x], y + half_y ]

    pos[:bottom][:left] = [ x, y + options[:cell][:size][:y] ]
    pos[:bottom][:center] = [ x + half_y, y + options[:cell][:size][:y] ]
    pos[:bottom][:right] = [ x + options[:cell][:size][:x], y + options[:cell][:size][:y] ]

    return pos
  end


  def self.cell_svg( symbols, position, options )
    str = ''
    symbols.each do | instruction |
      case instruction[ 0 ]
        when 'line'
          keys = instruction
            .drop( 1 )
            .map { | c | 
                keys = c
                  .to_s.split( '__' )
                  .map { | c | c.to_sym }
            }

          x1, y1 = position[ keys[ 0 ][ 0 ] ][ keys[ 0 ][ 1 ] ]
          x2, y2 = position[ keys[ 1 ][ 0 ] ][ keys[ 1 ][ 1 ] ]

          s_width = options[:style][:line][:stroke][:width].to_s
          s_color = options[:style][:line][:stroke][:color].to_s
          s_opacity = options[:style][:line][:stroke][:opacity].to_s
          s_linecap = options[:style][:line][:stroke][:linecap].to_s

          str += "<line x1=\"#{x1}\" y1=\"#{y1}\" x2=\"#{x2}\" y2=\"#{y2}\" style=\"stroke-width: #{s_width};stroke: #{s_color};stroke-opacity: #{s_opacity};stroke-linecap: #{s_linecap};\" transform=\"matrix(1,0,0,1,0,0)\" />"

        when 'ellipse'
          keys = instruction
            .drop( 1 )
            .map { | c | 
                keys = c
                  .to_s.split( '__' )
                  .map { | c | c }
            }

          cx, cy = position[ keys[ 0 ][ 0 ].to_sym ][ keys[ 0 ][ 1 ].to_sym ]
          rx = options[:cell][:size][:x] / 2
          ry = options[:cell][:size][:y] / 2

          s_width = options[:style][:ellipse][:stroke][:width]
          s_color = options[:style][:ellipse][:stroke][:color]
          s_opacity = options[:style][:ellipse][:stroke][:opacity]
          s_linecap = options[:style][:ellipse][:stroke][:linecap]
          e_fill = options[:style][:ellipse][:fill]
          str += "<ellipse cx=\"#{cx}\" cy=\"#{cy}\" rx=\"#{rx}\" ry=\"#{ry}\" style=\"stroke-width: #{s_width};stroke: #{s_color};stroke-opacity: #{s_opacity};stroke-linecap: #{s_linecap};; fill: #{e_fill}\" transform=\"matrix(1,0,0,1,0,0)\" />"

        when 'rectangle'
          keys = instruction
            .drop( 1 )
            .map { | c | 
                keys = c
                  .to_s.split( '__' )
                  .map { | c | c }
          }

          x, y = position[ keys[ 0 ][ 0 ].to_sym ][ keys[ 0 ][ 1 ].to_sym ]
          r_fill = options[:style][:rectangle][:fill][:color]
          r_opacity = options[:style][:rectangle][:fill][:opacity]

          width = options[:cell][:size][:x]
          height = options[:cell][:size][:y]

          str += "<rect x=\"#{x}\" y=\"#{y}\" width=\"#{width}\" height=\"#{height}\" style=\"fill: #{r_fill}; fill-opacity: #{r_opacity}\" transform=\"matrix(1,0,0,1,0,0)\" />"
      end
    end
    return str
  end


  def self.generate( lines, options )
    svg = ''
    width = options[:canvas][:size][:x]
    height = options[:canvas][:size][:y]
    fill = options[:style][:canvas][:fill][:color]
    fill_opacity = options[:style][:canvas][:fill][:opacity].to_s

    svg += "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" xmlns:jfreesvg=\"http://www.jfree.org/jfreesvg/svg\" width=\"#{width}\" height=\"#{height}\" text-rendering=\"auto\" shape-rendering=\"auto\">
  <defs></defs>"
    svg += "<rect x=\"0\" y=\"0\" width=\"#{width}\" height=\"#{height}\" style=\"fill: #{fill}; fill-opacity: #{fill_opacity}\" transform=\"matrix(1,0,0,1,0,0)\" />"
    svg += lines.clone
    svg += "</svg>"
    return svg
  end

end
